mage_ai/streaming/sinks/influxdb.py

```diff
@@ -0,0 +1,95 @@
+import time
+from collections.abc import Iterable
+from dataclasses import dataclass
+from typing import Dict, List
+
+from influxdb_client import InfluxDBClient, WriteOptions, WritePrecision
+
+from mage_ai.shared.config import BaseConfig
+from mage_ai.streaming.constants import DEFAULT_BATCH_SIZE, DEFAULT_TIMEOUT_MS
+from mage_ai.streaming.sinks.base import BaseSink
+
+
+def get_timestamp_ms():
+    return int(time.time() * 1e3)
+
+
+@dataclass
+class InfluxDbConfig(BaseConfig):
+    url: str
+    token: str
+    org: str
+    bucket: str = "data"
+    measurement: str = "default"
+    batch_size: int = DEFAULT_BATCH_SIZE
+    timeout_ms: int = DEFAULT_TIMEOUT_MS
+
+
+class InfluxDbSink(BaseSink):
+    config_class = InfluxDbConfig
+
+    def init_client(self):
+        self._print("Start initializing writer.")
+        # Initialize influxdb client and write_api
+        self.client = InfluxDBClient(
+            url=self.config.url,
+            token=self.config.token,
+            org=self.config.org,
+        )
+        if self.config.batch_size > 0:
+            batch_size = self.config.batch_size
+        else:
+            batch_size = DEFAULT_BATCH_SIZE
+        if self.config.timeout_ms > 0:
+            timeout_ms = self.config.timeout_ms
+        else:
+            timeout_ms = DEFAULT_TIMEOUT_MS
+
+        options = WriteOptions(
+            batch_size=batch_size,
+            flush_interval=timeout_ms,
+            jitter_interval=0,
+            retry_interval=5_000,
+            max_retries=5,
+            max_retry_delay=125_000,
+            exponential_base=2,
+        )
+        self.write_api = self.client.write_api(write_options=options)
+        self._print("Finish initializing writer.")
+
+    def write(self, data: Dict):
+        if isinstance(data, dict):
+            record = {
+                "measurement": data.get("measurement", self.config.measurement),
+                "time": data.get("timestamp", get_timestamp_ms()),
+                "tags": data.get("tags", {}),
+                "fields": data.get("fields", data),
+            }
+        elif isinstance(data, Iterable):
+            for i, value in enumerate(data):
```

Comment:
Do you only get the last record?

Reply:
I am assuming you refer to line 66: It's a bit tricky, not only the last record is taken but the whole record.

- If data["fields"] is found we assume the data can be found there.
- If data["fields"] does not exist, the record is malformed. But there is a chance that the data record is the data itself (but without "fields" and most probably without "tags", "time", "measurement") because the schema was ignored. Then we just take the whole data record as fields. It's not good practice to write everything as fields in InfluxDB, but it works.

Alternatively we could reject dict records, if the schema does not match.